rbw: add module

ambroisie opened the pull request:
### Description

`rbw` is a Bitwarden client written in `rust`. 

### Checklist

- [X] Change is backwards compatible.

- [X] Code formatted with `./format`.

- [X] Code tested through `nix-shell --pure tests -A run.all`.

- [X] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [X] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [X] Added myself and the module files to `.github/CODEOWNERS`.

berbiche commented on modules/programs/rbw.nix:
```diff
@@ -0,0 +1,95 @@
+        }
+      '';
+      description = ''
+        <command>rbw</command> configuration, if not defined the
+        configuration will not be managed by <command>home-manager</command>.
```
```suggestion
        rbw configuration, if not defined the
        configuration will not be managed by Home Manager.
```

Since you talking about projects, you don't need to add a `<command>` tag.

berbiche reviewed (changes requested):
Thanks for the contribution!

sumnerevans commented on modules/programs/rbw.nix:
```diff
@@ -0,0 +1,95 @@
+          '';
+        };
+
+        pinentry = mkOption {
+          type = types.package;
+          default = pkgs.pinentry;
+          defaultText = literalExample "pkgs.pinentry";
+          example = literalExample "pkgs.pinentry-gtk2";
+          description = ''
+            The pinentry program to use for password prompts.
+          '';
+        };
```
I wonder if it's worth unifying the pinentry config at some point. There's a lot of applications that rely on it (or may rely on it), so it doesn't make much sense that it's really only in the GPG agent config right now.

ambroisie replied:
Sure, I'll copy-paste it

berbiche replied:
> I wonder if it's worth unifying the pinentry config at some point. There's a lot of applications that rely on it (or may rely on it), so it doesn't make much sense that it's really only in the GPG agent config right now.

I was thinking the same thing but I don't know where it would fit.

ambroisie commented:
I had forgotten to add the news entry, it should be good now.

berbiche commented on modules/programs/rbw.nix:
```diff
@@ -0,0 +1,105 @@
+        {
+          email = "name@example.com";
+          lock_timeout = 300;
+          pinentry = pkgs.pinentry-gtk2;
```
This would need to be updated

berbiche commented on tests/modules/programs/rbw/settings.nix:
```diff
@@ -0,0 +1,33 @@
+{ pkgs, ... }:
+let
+  expectedRegex = ''
+    {
+      "base_url": "bitwarden.example.com",
+      "email": "name@example.com",
+      "identity_url": "identity.example.com",
+      "lock_timeout": 300,
+      "pinentry": "/nix/store/.*-pinentry"
+    }
+  '';
+in {
+  config = {
+    programs.rbw = {
+      enable = true;
+      settings = {
+        email = "name@example.com";
+        base_url = "bitwarden.example.com";
+        identity_url = "identity.example.com";
+        lock_timeout = 300;
+        pinentryFlavor = "tty";
+      };
+    };
+
+    nixpkgs.overlays = [ (import ./overlay.nix) ];
+
+    nmt.script = ''
+      assertFileRegex \
+        home-files/.config/rbw/config.json \
+        '${expectedRegex}'
+    '';
+  };
+}
```
Can you add an overlay for the selected pinentry method?

Overlay:
```
self: super: { pinentry.some-flavor = self.writeScriptBin "pinentry-some-flavor" '' // { outPath = "@pinentry@"; }; }
```

Then you can use `assertFileContent` instead of `assertFileRegex`:

```
expectedContent = ''
  stuff above
  "pinentry": "@pinentry@"
'';
```

ambroisie replied:
Okay I figured a way around it.

Also, the option must be named `pinentry` in the configuration file, not `pinentryFlavor` as I had pasted by error.

berbiche commented on modules/programs/rbw.nix:
```diff
@@ -0,0 +1,105 @@
+    };
+
+    settings = mkOption {
+      type = types.nullOr settingsModule;
+      default = null;
```
Personal preference: instead of making this type nullable, set the default to the empty attribute set `{}` and update the check below.

ambroisie replied:
Turns out this does not play well with the fact that `mail` is not null-able. The configuration is always being generated.

I'll keep it null-able instead.

berbiche commented on tests/modules/programs/rbw/settings.nix:
```diff
@@ -0,0 +1,34 @@
+
+    nmt.script = ''
+      assertFileExists home-files/.config/rbw/config.json
+      assertFileContains \
+        home-files/.config/rbw/config.json \
+        '${expected}'
```
```suggestion
      assertFileContent \
        home-files/.config/rbw/config.json \
        '${expected}'
```

Sorry, I meant `assertFileContent` yesterday.

berbiche commented on tests/modules/programs/rbw/simple-settings.nix:
```diff
@@ -0,0 +1,28 @@
+
+    nmt.script = ''
+      assertFileExists home-files/.config/rbw/config.json
+      assertFileContains \
+        home-files/.config/rbw/config.json \
+        '${expected}'
+    '';
```
```suggestion
      assertFileContent \
        home-files/.config/rbw/config.json \
        '${expected}'
    '';
```

berbiche commented on modules/programs/rbw.nix:
```diff
@@ -0,0 +1,105 @@
+        }
+      '';
+      description = ''
+        <command>rbw</command> configuration, if not defined the
+        configuration will not be managed by Home Manager.
```
```suggestion
        rbw configuration, if not defined the
        configuration will not be managed by Home Manager.
```

marsam commented on modules/programs/rbw.nix:
```diff
@@ -0,0 +1,105 @@
+
+    # Only manage configuration if not empty
+    (lib.mkIf (cfg.settings != null) {
+      xdg.configFile."rbw/config.json".source =
```
You might want to add a note that `settings` won't support darwin, since rbw uses `~/Library/Application Support/rbw/config.json` there

sumnerevans replied:
Even better would be to write to the correct file depending on the platform. See https://github.com/nix-community/home-manager/blob/master/modules/programs/lazygit.nix#L46 for an example.

sumnerevans commented on modules/programs/rbw.nix:
```diff
@@ -0,0 +1,112 @@
+        {
+          email = "name@example.com";
+          lock_timeout = 300;
+          pinentry = "gnome3";
```
The settings in the tests that you use below use "gnome" instead of "gnome3". I think that upstream is changing everything to be just "gnome" instead of "gnome3" in light of the GNOME 40 release?

```suggestion
          pinentry = "gnome";
```

(Same comment above in the pinentry description as well.)

ambroisie replied:
I use `gnome` in the tests because that's what the attribute name is in the overlay with the dummy packages.

As for the option description, I simply copy pasted the `pinentryFlavor` one from `gpg-agent`. Looking at `nixpkgs` it does seem to still be using `gnome3`.

sumnerevans reviewed (commented):
Is it supposed to be "gnome" or "gnome3"?

ambroisie commented:
@sumnerevans I used `gnome3` in the overlay to avoid confusion.

berbiche reviewed (approved):
LGTM!

Thanks again for the PR.

marsam commented on modules/programs/rbw.nix:
```diff
@@ -0,0 +1,112 @@
+        };
+
+        pinentry = mkOption {
+          type = types.enum pkgs.pinentry.flavors;
```
Unfortunately, this would make it hard to use with `pinentry_mac`
in darwin we usually use `${pkgs.pinentry_mac}/Applications/pinentry-mac.app/Contents/MacOS/pinentry-mac`

sumnerevans commented on modules/misc/news.nix:
```diff
@@ -2031,6 +2031,7 @@ in
       }
 
       {
+
```
nit: remove extraneous newline

sumnerevans commented:
I tested this out and it looks good. As far as the test failure, you should be able to rebase on `master` now and it should pass. Also, please address the comment from @marsam. Thanks!

ambroisie commented:
Sorry, I thought this had been merged :sweat_smile: 

I rebased and removed the merge into the branch (I find them untidy).

As for the pinentry, the `gpg-agent` service does not allow using another derivation for its pinentry (even though it is not gated behind `stdenv.isLinux`), I assume any solution done here should be reflected there too?

sumnerevans commented:
> As for the pinentry, the `gpg-agent` service does not allow using another derivation for its pinentry (even though it is not gated behind `stdenv.isLinux`), I assume any solution done here should be reflected there too?

Good point. We can defer that work for a different PR.
